### PR TITLE
Add EventedSet class

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This adds an `EventedSet` class.  just like `EventedList`, but for sets :)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [x]  the test suite for my feature covers event emission and parity with `builtin.set`
- [x] all tests pass with my change

### CODE REVIEW

File: napari/utils/events/containers/_set.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        self._set: set[_T] = set()
+        self.update(data)
+
+    # >>>>>>>> Required Abstract Methods
+
+    def __contains__(self, x: Any) -> bool:
+        return x in self._set
+
+    def __iter__(self) -> Iterator[_T]:
+        return self._set.__iter__()
```
Comment (Carreau) [inline napari/utils/events/containers/_set.py]:
```suggestion
        return iter(self._set)
```

File: napari/utils/events/containers/_set.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    def discard(self, value: _T) -> None:
+        """Remove an element from a set if it is a member.
+
+        If the element is not a member, do nothing.
+        """
+        if value in self:
+            self._set.discard(value)
+            self.events.removed(value=value)
+
+    # <<<<<<<< End Abstract Methods
```
Comment (Carreau) [inline napari/utils/events/containers/_set.py]: Do those marker have any change of confusing tools that deal with git conflicts ?

File: napari/utils/events/containers/_set.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        Elements to initialize the set with.
+
+    Events
+    ------
+    added (value: _T)
+        emitted after an item is added to the set.  Will not be emitted if an
+        item was already in the set when added.
+    removed (value: _T)
+        emitted after an item is removed from the set.  Will not be emitted if
+        the item was not in the set when discarded.
```
Comment (Carreau) [inline napari/utils/events/containers/_set.py]: do we want to have different event for batch changes in particular `update()` ? Or even have `added` have a set of new values ? doing one event per insertion in particular in `update()` seem at risk of triggering accidental O(N^2) behavior; like if a gui element start to do N repainting on a set update.

**Review by Carreau — APPROVED** [review]
+1, one comment on batch operation where I'm slightly concern about accidental bad complexity and wondering if the events should contain many values. 

I would understand if it's uniformity worth other containers, but that seem to not be the case as list seem to emit even after and before changes ?

Still if you don't have concern about the complexity I don't think there is anything delaying this.

File: napari/utils/events/containers/_set.py (hunk 2 above)
Comment (tlambert03) [reply napari/utils/events/containers/_set.py]: oh!  they might! :joy: ... will change

File: napari/utils/events/containers/_set.py (hunk 3 above)
Comment (tlambert03) [reply napari/utils/events/containers/_set.py]: sure sounds good.  let's stick to one event name `added` ... do you think it should _always_ emit an iterable value, even it adds a single value, for the sake of simplicity on the receiving end?

Comment (Carreau) [reply napari/utils/events/containers/_set.py]: I think that yes, always an iterable should be good. That is what I would have done.

If you want you could even send a single change event with two added and remove attributes, which might make it easier to only hook one callback in general and avoid two events in `symmetric_difference_update`; I'm of two mind with this alternative though.

Comment (tlambert03) [reply napari/utils/events/containers/_set.py]: updated to minimize events.  all events now have a `value` of type `set`

**Review by sofroniewn — APPROVED** [review]
Love it, looks good to me!! Will be nice to have for if we want to make `selection` on a list an evented set. Thanks very much for adding!!

**Pull request merged**